### PR TITLE
Rendering parse error context #205

### PR DESCRIPTION
**Pull request by recursion-ninja**
An *initial* implementation to render the line of a parse error along with the parse error. Addresses issue #205.

The bar on the side is intended to mimic GHC's 8.2.0's new contextual warning/error rendering.

Let me know your feedback .

Some example renderings:
```
*ghci> let parseTest2 = (\p s -> either (putStr . parseErrorPrettyWithLine s) print $ parse p "example.txt" s) :: Show a => Parsec Void String a -> String -> IO ()
*ghci> parseTest2 (count 10 numberChar) "123456789"
example.txt:1:10:
   |
 1 | 123456789
   |          ^
unexpected end of input
expecting numeric character
*ghci> parseTest2 (count 10 (numberChar <|> Text.Megaparsec.Char.newline)) "12\n3\n45y"
example.txt:3:3:
   |
 3 | 45y
   |   ^
unexpected 'y'
expecting newline or numeric character
*ghci> parseTest2 (count 10 (numberChar <|> Text.Megaparsec.Char.newline)) "12\n3\n45\n"
example.txt:4:1:
   |
 4 | <empty line>
   | ^
unexpected end of input
expecting newline or numeric character
*ghci> putStrLn "Note the contextual line number padding on the left-hand side:"
Note the contextual line number padding on the left-hand side:
*ghci> parseTest2 (count 10010 (numberChar<|>Text.Megaparsec.Char.newline)) $ replicate 10000 '\n' <> "123456789"
example.txt:10001:10:
       |
 10001 | 123456789
       |          ^
unexpected end of input
expecting newline or numeric character
```

### CODE REVIEW

**Comment by recursion-ninja** [comment]
This still needs to be tested with a stack of source positions that has a depth greater than one.

We might also want to consider when a parser error contains multiple unexpected tokens, underlining each unexpected token that is on the same line as the first unexpected token.

File: Text/Megaparsec/Error.hs [hunk 1]
```diff
@@ -248,6 +252,98 @@ parseErrorPretty
 parseErrorPretty e =
   sourcePosStackPretty (errorPos e) <> ":\n" <> parseErrorTextPretty e
 
+
+-- | Pretty-print a 'ParseError' and display the line on which the parse error
+-- occurred. The rendered 'String' always ends with a newline.
+--
+-- Assumes that the stream tokens can be coerced to an @Int@ and that the value
```
Comment (mrkkrp) [inline Text/Megaparsec/Error.hs]: I think "converted" is a better word to use here, as "coerced" implies that we use actual coercion between types that is supported by GHC (`coerce` for type-safe coercion and `unsafeCoerce` for unsafe coercion), which is not the case.

You could also mention `fromEnum`, so it's clear how the "conversion" is performed.

File: Text/Megaparsec/Error.hs [hunk 2]
```diff
@@ -248,6 +252,98 @@ parseErrorPretty
...
+streamIter f p ts = fst $ go (x-1) (f ts)
+  where
+    x = unPos p
+    go 0 val    = val
+    go n (_,xs) = let val = f xs
+                  in go (n-1) val
+
+-- | Takes a line from a stream.
+--
+-- Tokens which are numericaly coercable to the integer value @10@ are considered
```
Comment (mrkkrp) [inline Text/Megaparsec/Error.hs]: Again here "coercable" -> "convertable".

File: Text/Megaparsec/Error.hs [hunk 3]
```diff
@@ -248,6 +252,98 @@ parseErrorPretty
...
+grabLine ts =
+  -- remove the newline from the stream
+  case take1_ ts' of
+      Nothing    -> (line, ts')
+      Just (_,v) -> (line, v  )
+  where
+    -- grab the line content
+    (line, ts') = takeWhile_ isNotNewLine ts
+
+    -- Maybe instead we want to corece to an @Int@, then convert to a @Char@,
```
Comment (mrkkrp) [inline Text/Megaparsec/Error.hs]: No, that'll be most certainly slower, and `defaultAdvance1` is very sensitive to slow-downs because it's called on every token. Comparison with plain number 10 is OK.

File: Text/Megaparsec/Error.hs [hunk 4]
```diff
@@ -248,6 +252,98 @@ parseErrorPretty
...
+    cs =
+      case ts of
+        [] -> "<empty line>"
+        xs -> chr . fromEnum <$> xs
+    ts = chunkToTokens (Proxy :: Proxy s) lineVal
+    lineVal = streamIter grabLine linePos s
+    linePos = sourceLine p
+    columnNum = unPos $ sourceColumn p
+
+-- | Applies a "taking" function @n@ number of times to a 'Stream' and returns
```
Comment (mrkkrp) [inline Text/Megaparsec/Error.hs]: You need to escape the double quote character even though this Haddock is not public, it's a good idea to write correct Haddock markup anyway:

> Applies a \"taking\" ...

And there is a typo on the second like in `"tking"`.

File: Text/Megaparsec/Error.hs [hunk 5]
```diff
@@ -248,6 +252,98 @@ parseErrorPretty
...
+    lineNumberStr = show $ unPos linePos
+    padLength = length lineNumberStr + 2
+    padding = replicate padLength ' '
+
+    cs =
+      case ts of
+        [] -> "<empty line>"
+        xs -> chr . fromEnum <$> xs
+    ts = chunkToTokens (Proxy :: Proxy s) lineVal
+    lineVal = streamIter grabLine linePos s
```
Comment (mrkkrp) [inline Text/Megaparsec/Error.hs]: I see `streamIter` and `grabLine` are only used once here, for dropping `linePos` lines from the beginning of the stream. I would like you to merge the both functions into a single function that just clearly does what we need here: dropping N lines and returning the line we arrive at after that. This will be more clear to future readers of the code.

File: Text/Megaparsec/Error.hs [hunk 6]
```diff
@@ -248,6 +252,98 @@ parseErrorPretty
...
+  :: ( Enum (Token s)
+     , Stream s)
+  => (s -> (Tokens s, s)) -- ^ A function which takes part of the stream
+  -> Pos                  -- ^ Number of times to apply the function.
+  -> s                    -- ^ Stream to be taken from
+  -> Tokens s             -- ^ The the last segment taken from the stream
+streamIter f p ts = fst $ go (x-1) (f ts)
+  where
+    x = unPos p
+    go 0 val    = val
```
Comment (mrkkrp) [inline Text/Megaparsec/Error.hs]: Don't forget to add strictness annotation to the both arguments.

File: Text/Megaparsec/Error.hs [hunk 7]
```diff
@@ -248,6 +252,98 @@ parseErrorPretty
...
+    paddingLine   = padding <> "|"
+    contextLine   = " " <> lineNumberStr <> " | " <> cs
+    pointingLine  = padding <> "|" <> replicate columnNum ' ' <> "^"
+
+    lineNumberStr = show $ unPos linePos
+    padLength = length lineNumberStr + 2
+    padding = replicate padLength ' '
+
+    cs =
+      case ts of
```
Comment (mrkkrp) [inline Text/Megaparsec/Error.hs]: Two things I'm not sure about:

1. Is printing `"<empty line>"` better than just printing empty line?
2. How to handle the cases when we point to a `'\n'` character? Probably we should place the caret `^` to the end of line the `'\n'` character ends? This is a corner case we should test.

File: Text/Megaparsec/Error.hs [hunk 8]
```diff
@@ -248,6 +252,98 @@ parseErrorPretty
...
+  => (s -> (Tokens s, s)) -- ^ A function which takes part of the stream
+  -> Pos                  -- ^ Number of times to apply the function.
+  -> s                    -- ^ Stream to be taken from
+  -> Tokens s             -- ^ The the last segment taken from the stream
+streamIter f p ts = fst $ go (x-1) (f ts)
+  where
+    x = unPos p
+    go 0 val    = val
+    go n (_,xs) = let val = f xs
+                  in go (n-1) val
```
Comment (mrkkrp) [inline Text/Megaparsec/Error.hs]: I'm not entirely sure why this can't be written just as `go (n-1) (f xs)`, but this is not really important.

**Comment by mrkkrp** [comment]
Don't mind Travis, it has a weird spread of build times ranging from 20 minutes or less to 50 minutes where the process gets killed.

**Comment by mrkkrp** [comment]
> We might also want to consider when a parser error contains multiple unexpected tokens

This is no longer possible in version 6. I'll write a blog post about that, but in short we always have one unexpected token at the moment of parse error generation (makess sense, you consider "unexpected" what you run into, and you can't run into two different things at the same position at once) and several unexpected tokens only happen because of merging of different parse errors from different branches of parsing. I had to change the merging algorithm to make parse errors generated by the new `tokens` function merge more nicely and so some normalization were required. That normalization however prevents us from having `ParseError` as lawful monoid (which we would like very much) unless we switch from set of unexpected tokens to `Maybe`.

**Comment by recursion-ninja** [comment]
Thanks for all the feedback. I'll address all the issues you made note of.

I ran my branch on a bio-informatics input file that I placed a few parse errors in and the rendering was less than ideal. The input file contains segments of a genome and has lines within the file that are 10,000 characters long. Rendering out the whole 10,000 character line isn't terribly informative so I thought that we could pass in a "token range" parameter that says, render at most this many tokens around the line of the parse error. A range of 80 or 120 should be appropriate for most files, but the user can specify the token range manually if they want. 

I think this can generalize to work on a stream of arbitrary tokens, not just with "numerically-Char-like" tokens. Let me know what you think. I'll have time to try implementing it in about 12 hours.

**Comment by mrkkrp** [comment]
Yeah "token range" is a good idea.

**Comment by mrkkrp** [comment]
Any progress on this?

**Comment by recursion-ninja** [comment]
I'm about halfway through adding the "token range." I added a "total-tokens-consumed" field to the parse error to make things easier in the error message post-processing with the stream in context.

**Comment by mrkkrp** [comment]
>  I added a "total-tokens-consumed" field to the parse error to make things easier

We can't go that way, IMO. The field is essentially useless because the total number of processed tokens is an accumulated value from entire parser run, while this use case demands something tied to an individual file. The difference is apparent when we have include files or something like that.

**Comment by mrkkrp** [comment]
I'm planning to push towards release of version 6 next week. There are also a couple of libraries I want to upgrade to Megaparsec 6 soon and a new project that requires it. Would it be possible for you to push what you have so far and I could pick up your ideas and finish the PR? That's in case you don't have time to finish yourself in a couple of days.

Sorry to hurry you :-|

**Comment by recursion-ninja** [comment]
No worries about "hurrying" me. I think I got the core contextualized rendering complete. I haven't had sufficient time to test all the corner cases yet nor address some of the code review comment you made earlier. I pushed what I have so far.

Hopefully I can add a test suite and remove any defects over the next week, along with improve the documentation and address all code review comments had/have. Feel free to move forward with what I have so far.

**Comment by recursion-ninja** [comment]
> We can't go that way, IMO. The field is essentially useless because the total number of processed tokens is an accumulated value from entire parser run, while this use case demands something tied to an individual file. The difference is apparent when we have include files or something like that.

Do you have an example somewhere demonstrating this kind of parsing case? I'm not sure I follow, and would like to poke around with an example if possible so I can better understand the issue.

**Comment by mrkkrp** [comment]
Imagine you parse file A and you have processed 10 tokens from there. Next you run into an include pragma and so you set new input (file B) and use [`pushPosition`](https://hackage.haskell.org/package/megaparsec-5.3.1/docs/Text-Megaparsec-Prim.html#v:pushPosition) to set position at the beginning of the file. The counter for the total number of processed tokens isn't reset however, so if at this point a parse error occurs in file B, you'll have greater number of processed tokens than you expect (greater by 10).

**Comment by mrkkrp** [comment]
I'll start working on finishing this later today (notifying you so we don't duplicate effort). I'll add tests too.

**Comment by mrkkrp** [comment]
To be continued in #236.

**Comment by mrkkrp** [comment]
@recursion-ninja, How would you like to be added to the `AUTHORS.md` file?

**Comment by recursion-ninja** [comment]
```Alex Washburn (@recursion-ninja)```
